internal/reflect/struct.go

```diff
@@ -234,8 +234,18 @@ func FromStruct(ctx context.Context, typ attr.TypeWithAttributeTypes, val reflec
 			}
 		}
 
+		tfObjTyp := tfObjVal.Type()
+
+		// TODO: improve this test case
+		// TODO: add this to slices
+		// If the original attribute type is tftypes.DynamicPseudoType, the value is a concrete type (like tftypes.String, tftypes.List, etc.)
+		// however, the type used to build the tftypes.Object needs to be tftypes.DynamicPseudoType.
+		if attrTypes[name].TerraformType(ctx).Is(tftypes.DynamicPseudoType) {
+			tfObjTyp = tftypes.DynamicPseudoType
+		}
+
 		objValues[name] = tfObjVal
-		objTypes[name] = tfObjVal.Type()
+		objTypes[name] = tfObjTyp
```

Comment:
Another pessimistic change to avoid refactoring more of this code path.

I believe we could technically just use `objTypes[name] = attrTypes[name].TerraformType(ctx)` for everything, but wanted to make a non-invasive change 😄